### PR TITLE
[wip] build: Rebuild when a go files gets deleted

### PR DESCRIPTION
**Pull request by aequitas**
Fixes: https://github.com/purpleidea/mgmt/issues/347

Not tested to full extent but this should fix the problem of deleted files.

### CODE REVIEW

**Comment by aequitas** [comment]
@purpleidea other thing i noticed, skipping `vendor/` shaves of a few couple of thousands of nanoseconds from make. Which is valuable for when tab-completion targets need to be calculated. I was thinking maybe adding a flag to allow vendor files to be evaluated, or to just require `touch main.go; make` as working in vendor files is not that common. Whats your thought?

With:
```
$ time make
Generating: bindata...
Generating: lang...
/Applications/Xcode.app/Contents/Developer/usr/bin/make --quiet -C lang
        1.29 real         0.98 user         0.17 sys
```

Without:
```
$ time make
Generating: bindata...
Generating: lang...
/Applications/Xcode.app/Contents/Developer/usr/bin/make --quiet -C lang
        0.49 real         0.23 user         0.14 sys
```

**Comment by purpleidea** [comment]
You are brilliant, but I'm a bit afraid to merge this...

File: Makefile [hunk 1]
```diff
@@ -124,7 +125,13 @@ lang:
 $(PROGRAM): build/mgmt-${GOHOSTOS}-${GOHOSTARCH}
 	cp $< $@
 
-$(PROGRAM).static: $(GO_FILES)
+# keep a list of all go files so we can rebuild if one gets deleted
+.make.gofiles: $(GO_FILES) FORCE
+	@echo $^ > $@.tmp
+	@diff $@.tmp $@ >/dev/null || mv $@.tmp $@
+FORCE: ;
```
Comment (purpleidea) [inline Makefile]: Is there a need for this even if we can make clean && make build ?

**Comment by purpleidea** [comment]
Small issue:

When testing this by running `make` I get:

```
make: execvp: /usr/bin/env: Argument list too long
make: *** [Makefile:130: .make.gofiles] Error 127
```

**Comment by aequitas** [comment]
@purpleidea can you test with the latest update?

**Comment by purpleidea** [comment]
I still get the same issue. I think the problem is:

```
	@echo "$^" > $@.tmp 
```

because that list of files is too long for echo. Probably needs to be done in a loop or some trickery so that it doesn't hit the bash limit.

**Comment by aequitas** [comment]
You probably have a wee bit more files in your repo than I have atm :D 

I tried with https://www.gnu.org/software/make/manual/html_node/File-Function.html but somehow that doesn't work on my machine. Maybe due to an older make version. I'll do some more research.

**Pull request closed without merging**